### PR TITLE
Core/Quest: Handle quest 29409 (wandering isle)

### PR DESCRIPTION
**Pull request by aquadeus**
**Changes proposed:**

-  Handle Jaomin Ro fight more blizzlike

**Tests performed:**

Tested in-game

### CODE REVIEW

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 1]
```diff
@@ -400,10 +403,231 @@ struct npc_instructor_zhi : public ScriptedAI
...
+            {
+                me->GetMotionMaster()->MoveTargetedHome();
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
+    void DamageTaken(Unit* attacker, uint32& damage, DamageEffectType /*damageType*/, SpellInfo const* /*spellInfo = nullptr*/) override
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:496]: use OnHealthDepleted instead and UNKILLABLE static flag

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 2]
```diff
@@ -400,10 +403,231 @@ struct npc_instructor_zhi : public ScriptedAI
...
+
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.ScheduleEvent(EVENT_RANDOM_SPELL, 4s, 6s);
+    }
+
+    void JustReachedHome() override
+    {
+        if (me->HasUnitFlag(UNIT_FLAG_NON_ATTACKABLE_2))
+            me->RemoveUnitFlag(UNIT_FLAG_NON_ATTACKABLE_2);
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:447]: Don't both with HasFlag before RemoveFlag if there is no other code behind this condition

File: sql/updates/world/master/2024_99_99_99_quest_29409.sql [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+INSERT INTO `creature_text` (`CreatureID`, `GroupID`, `ID`, `Text`, `Type`, `Language`, `Probability`, `Emote`, `Duration`, `Sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
+(54611, 0, 0, 'A challenger approaches.', 12, 0, 100, 113, 0, 0, 59692, 0, 'Jaomin Ro to Player'),
+(54611, 1, 0, 'You''ve been trained well.', 12, 0, 100, 2, 0, 0, 55205, 0, 'Jaomin Ro to Player'),
+(54611, 1, 1, 'Well fought. Master Shang will be pleased.', 12, 0, 100, 2, 0, 0, 55206, 0, 'Jaomin Ro to Player'),
+(54611, 1, 2, 'Amazing! You''re much stronger than you look.', 12, 0, 100, 2, 0, 0, 55207, 0, 'Jaomin Ro to Player'),
+(54611, 1, 3, 'It seems I have more training to do. Thank you for the lesson.', 12, 0, 100, 2, 0, 0, 55208, 0, 'Jaomin Ro to Player'),
+(54611, 1, 4, 'I bow to your skill. You bested me fairly.', 12, 0, 100, 2, 0, 0, 55209, 0, 'Jaomin Ro to Player'),
+(54611, 1, 5, 'It was an honor to be defeated by you.', 12, 0, 100, 2, 0, 0, 55210, 0, 'Jaomin Ro to Player');
+
+UPDATE `creature_template_difficulty` SET `StaticFlags1`=268435456, `VerifiedBuild`=53441 WHERE (`Entry`=54611 AND `DifficultyID`=0); -- Jaomin Ro
```
Comment (Shauren) [inline sql/updates/world/master/2024_99_99_99_quest_29409.sql:11]: StaticFlags1 = 0x1008000E (NO_XP, NO_LOOT, UNKILLABLE, AMPHIBIOUS, CAN_SWIM)
StaticFlags3 = 0x00800000 (CAN_BE_MULTITAPPED - NYI, but not relevant anymore)
StaticFlags4 = 0x80000000 (QUEST_BOSS, same as type_flags)

File: sql/updates/world/master/2024_99_99_99_quest_29409.sql [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+UPDATE `creature_template_difficulty` SET `ContentTuningID`=80, `VerifiedBuild`=53441 WHERE (`Entry`=57750 AND `DifficultyID`=0); -- Jaomin Ro
```
Comment (Shauren) [inline sql/updates/world/master/2024_99_99_99_quest_29409.sql:12]: StaticFlags1 = 0x80206 (NO_XP, NO_LOOT, UNINTERACTIBLE, AMPHIBIOUS)

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 5]
```diff
@@ -17,13 +17,16 @@
...
+#include "CreatureAIImpl.h" // for RAND()
 #include "GridNotifiersImpl.h"
 #include "MotionMaster.h"
 #include "ObjectAccessor.h"
 #include "Player.h"
 #include "ScriptedCreature.h"
 #include "ScriptMgr.h"
+#include "SpellScript.h"
 #include "TaskScheduler.h"
+#include "Unit.h"
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:29]: Unneccessary, Player already includes Unit

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 6]
```diff
@@ -400,10 +403,231 @@ struct npc_instructor_zhi : public ScriptedAI
...
+        if (me->HealthBelowPctDamaged(2, damage))
+        {
+            damage = 0;
+
+            me->SetReactState(REACT_PASSIVE);
+            me->AttackStop();
+            me->RemoveAllAuras();
+            me->InterruptNonMeleeSpells(true);
+            _events.Reset();
+            me->SetUnitFlag(UNIT_FLAG_UNINTERACTIBLE | UNIT_FLAG_NON_ATTACKABLE_2);
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:507]: Use `Unit::SetUninteractible` to change UNIT_FLAG_UNINTERACTIBLE

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 7]
```diff
@@ -400,10 +403,231 @@ struct npc_instructor_zhi : public ScriptedAI
...
+    void JustAppeared() override
+    {
+        me->SetReactState(REACT_PASSIVE);
+        me->SetSpeedRate(MOVE_RUN, 2.5f);
+    }
+
+    void IsSummonedBy(WorldObject* summoner) override
+    {
+        DoCast(SPELL_FORCE_SUMMONER_TO_RIDE);
+        orientation = me->GetAbsoluteAngle(summoner->ToUnit()->GetVictim()) - me->GetOrientation();
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:543]: Check if summoner is a unit and that it has a target

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by aquadeus** [comment]
do i have to set isKill false in OnHealthDepleted?

**Comment by Shauren** [comment]
No, its not a value you set for outside code, its information for you to know if this hook is called on reaching 1 hp with UNKILLABLE or actually dying (its called in both cases)

**Comment by aquadeus** [comment]
oh ok ty 😃

**Pull request merged**